Registered fd support

DXist opened the pull request:
This is initial implementation of registered file descriptors - tested only on IO uring.

A pool of descriptors is registered when driver is created. For each operation that creates a file descriptor users provide the chosen `RegisteredFd` values and driver issues synchronous `register_files_update`.

For Windows driver emulates registration and maps registered fd to raw fds during interaction with kerhel.

I changed File's open/create method to async because it's possible to open files asyncronously with IO uring.

Later it's possible to implement async OpenAt and Socket operations that could create a direct registered file descriptor.

Berrysoft commented:
Please do not change the signature of `File::open` etc. to accept `RegisteredFd`. That's terrible. `RegisteredFd` is a low-level implementation detail. Instead, let the runtime or the driver allocate the registered fd.


It's OK to change `open`, `connect`, `bind` etc. to async, and you don't need to think about IOCP and mio.

Just leaving registered fd feature as io-uring specific is OK.

DXist commented:
It seems that bind & listen for direct socket descriptors aren't supported yet - https://github.com/axboe/liburing/issues/941

These syscalls don't do any IO and are issued not often. Probably doesn't make sense to provide async API.

DXist commented:
@Berrysoft , I've added logic that tracks registered file descriptors in boxed bit slice. Tests on IO uring pass except the doctests related to async open operation.

IOCP port is configured inside registration logic so I removed `attach` method.

Berrysoft commented:
Would you resolve the conflicts? There are many changes merged recently.

DXist commented:
Ok. Do you think it makes sense to provide non async File methods when runtime feature is not enabled?

Berrysoft commented:
I prefer a unified interface. If only one  target needs async, make the method async on all platforms and add comments about that. (see sync_* in File.)

DXist commented:
I reverted to sync open for unified interface and integrated RegisteredFd with mio.

Also set blocking socket for Linux and fixed tcp_connect tests.

Berrysoft commented on README.md:
```diff
@@ -45,18 +45,19 @@ While you can also control the low-level driver manually:
 use compio::{
     buf::IntoInner,
-    driver::{AsRawFd, Driver, Poller},
+    driver::{AsRawFd, AsRegisteredFd, Driver, Poller, RegisteredFileDescriptors},
```
I think we need to talk about the name "registered". It's too long. Maybe `RegFd` is OK.

Berrysoft commented on benches/fs.rs:
```diff
@@ -34,7 +34,7 @@ fn read(c: &mut Criterion) {
             use tokio::io::AsyncReadExt;
 
-            let mut file = tokio::fs::File::open("Cargo.toml").await.unwrap();
+            let mut file = tokio::fs::File::open("Cargo.toml").unwrap();
```
tokio needs await. Have you turned on your rust-analyzer?

Berrysoft commented on src/driver/iocp/mod.rs:
```diff
@@ -112,14 +116,21 @@ pub trait OpCode {
     ///
     /// `self` must be alive until the operation completes.
-    unsafe fn operate(&mut self, optr: *mut OVERLAPPED) -> Poll<io::Result<usize>>;
+    unsafe fn operate(
```
I don't agree with changing IOCP and mio driver to support reg fds. Let's make it io-uring specific.

DXist replied:
Do you mean that for IOCP/mio RegFd is a type alias for RawFd?

Berrysoft replied:
I mean that IOCP/mio should never know about RegFd. It shouldn't have interface to accept/return RegFd.

Berrysoft commented on src/driver/registered_fd.rs:
```diff
@@ -0,0 +1,216 @@
+pub const UNREGISTERED: RawFd = windows_sys::Win32::Foundation::INVALID_HANDLE_VALUE as RawFd;
+
+/// Public registration API
+pub trait RegisteredFileDescriptors: RegisteredFileAllocator {
```
Maybe we can call it `RegFds`

Berrysoft commented on src/driver/registered_fd.rs:
```diff
@@ -0,0 +1,216 @@
+// Platforms that don't have system provided file descriptor registry use
+// driver emulated one
+#[doc(hidden)]
+pub(super) trait FDRegistry {
```
At least it should be `FdRegistry`

Berrysoft commented on src/task/op.rs:
```diff
@@ -64,21 +64,24 @@ impl OpRuntime {
 
     pub fn update_waker<T>(&mut self, key: Key<T>, waker: Waker) {
-        if let Some(op) = self.ops.get_mut(*key) {
+        if let Some(op) = self.ops.get_mut(usize::try_from(*key).expect("in range")) {
```
You see, you changed the type of `user_data` to u64. It's right for io-uring, but here is more complex. Nowadays even the users of 32-bit Windows is decreasing, let alone the users of 32-bit Linux. I cannot imagine a user with 32-bit Linux and io-uring. I don't like this change.

DXist replied:
I want to pack structure into `user_data` to dispatch callbacks between multiple IO clients with a dedicated queue for each IO operation (known max IO depth). u64 gives fixed size.

So in general `user_data` could be used not as a index in global operations queue.

Berrysoft replied:
`user_data` is used a key when cancelling it (see `AsyncCancel` opcode). In `compio`, it is also used for `Token` in `mio`, and a hashmap index for IOCP. I'm here considering other platforms for you to make sure the public, cross-platform general interface is suitable for all different kind of drivers.

Berrysoft commented on tests/tcp_connect.rs:
```diff
@@ -45,9 +45,11 @@ test_connect_ip! {
-        assert!(listener.accept().await.is_ok());
+        let res = listener.accept().await;
+        if let Err(err) = res {
+            panic!("{}", err);
```
Then just `unwrap` is OK.

Berrysoft commented:
The error on Windows is `WSAEFAULT`, which means the sockaddr pointer or length is invalid.

https://learn.microsoft.com/en-us/windows/win32/winsock/windows-sockets-error-codes-2

Berrysoft commented:
In this PR, you chose to add registered file descriptor as a *cross-platform* feature. That's what I'm against to. RegFds should be only for io-uring. Neither IOCP nor mio will benefit from that.

However, if you just make the `RawFd` on Linux means the registered file descriptor, that's OK, because it won't influence the logic of other drivers.

Or, if you make the registered as a platform-dependent feature, it's also OK. When a user uses this feature, they should be aware that it's a Linux-only feature.

DXist commented:
Another option is to have another fd type for operations that support registered file descriptors. On mio/windows it could be a type alias for RawFd. For IO uring we could alias it to `io_uring::types::Target`. Or use own type wrapper and own `UseFixed` trait.

High level API (File, Socket, etc.) items are always created with raw fds. It's possible to provide optional registration/deregistration for advanced users targeting Linux.

So user will have control on file descriptor tables both on application side (open/close) and kernel side (register/unregister).

Berrysoft commented:
Why not define `RawFd` as an alias of `Target`? However, `Target` is a private type, so we may need to declare a new one.

DXist commented:
> Why not define `RawFd` as an alias of `Target` for io-uring? However, `Target` is a private type, so we may need to declare a new one.

Yes, it's inside a sealed module.

We can make a structure that always holds RawFd (for AsRawFd implementation always point to raw fd and optionally has RegFd. RegFd wraps Tokio's Fixed(u32) and non Send/Sync marker.

IO uring implementation either uses Fixed(u32) or creates Fd(RawFd).

Berrysoft commented:
What I was always saying is
``` rust
pub type RawFd = io_uring::types::Target;
```
-OR-
``` rust
pub enum RawFd {
    Fd(i32),
    Fixed(u32),
}
```
but you were always saying defining two types... Did you understand my comments above?

DXist commented:
We could do like that. For operations that are done via syscalls instead of IO uring we check if we got Fd(_) variant. If we got Fixed - it's a users mistake and we reveal it in runtime and panic.

DXist commented:
Can we have `Fd` type in operations? In mio/windows Fd is RawFd alias.

On Linux it's an enum 

```
pub enum Fd {
    Raw(RawFd),
    Fixed(RegFd),
}
```

Berrysoft commented:
The name `Fd` is OK.

George-Miao commented:
Closed for lack of activity. Re-open if you're interested in continuing contribution.
